[UXUI-225] Make values in form field properties get automatically generated

andersonjeccel opened the pull request:
## Description

This PR implements automatic value generation for form field properties in the SortableValueLabelListType. When a user enters a label but leaves the value field empty, the system will automatically generate a URL-safe slug from the label text.

Users no longer need to manually create values for form field properties


---
### 📋 Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally
2. Navigate to Forms, create one, add a Select field
3. Open the Properties tab and click to Add
4. Enter a label in the label field (e.g., "My Custom Option")
5. Leave the value field empty
6. Submit the form (save/add the field)
7. Verify that the value field is automatically populated with a slugified version when you open the field config again and check the Properties tab (e.g., "my_custom_option")

<img width="598" height="465" alt="image" src="https://github.com/user-attachments/assets/c19c975d-af24-44da-95f3-42e876f3b658" />

andersonjeccel committed:
Refactor SortableValueLabelListType and add unit test

andersonjeccel committed:
test: Rector

andersonjeccel committed:
test: phpstan

andersonjeccel committed:
test: csfix

andersonjeccel committed:
Update cs fixer version

andersonjeccel committed:
Improve quality

andersonjeccel committed:
Update value field label to 'Value (autogenerated)' and add translation string

andersonjeccel committed:
Add comprehensive test coverage for SortableValueLabelListType

andersonjeccel committed:
Revert "Update cs fixer version"

This reverts commit 620bba0fc7892c9189ccd4da193e6a17f075fc91.

andersonjeccel committed:
Improve code quality

andersonjeccel committed:
csfix

andersonjeccel committed:
rector

andersonjeccel committed:
Improve code quality

RCheesley committed:
Merge branch '7.x' into UXUI-225-make-values-in-form-field-properties-get-automatically-generated

RCheesley reviewed (changes requested):
I figured it might be important to test this with international characters so I used a list like this:

<img width="1920" height="934" alt="Edit Form - Testing the forms _ Mautic (2)" src="https://github.com/user-attachments/assets/5e36c4f5-949c-4f50-8dd0-86ce2c9fce93" />

When I saved, a bunch of the fields are completely missing:

<img width="1920" height="934" alt="Edit Form - Testing the forms _ Mautic (3)" src="https://github.com/user-attachments/assets/3967819a-9450-418a-afa8-55522a92a4e4" />

It looks like as soon as it gets to the Russian, Chinese, Japanese and Arabic characters, the slug is not created. The same is true for Greek and Korean.

I'm using DDEV locally for testing this.

Here are the strings I used for testing (thanks ChatGPT :) )

| Language      | Example Word(s)  |
|:--------------|:-----------------|
| French        | Numéro (Number), Âge (Age) |
| German        | Straße (Street), Über (Over) |
| Spanish       | Teléfono (Phone), Dirección (Address) |
| Portuguese    | Coração (Heart), Ação (Action) |
| Swedish       | Födelsedag (Birthday), Äktenskap (Marriage) |
| Russian       | Номер (Number), Адрес (Address) |
| Chinese       | 电话 (Phone), 地址 (Address) |
| Japanese      | 電話番号 (Phone Number), 住所 (Address) |
| Arabic        | رقم (Number), عنوان (Address) |
| Turkish       | Adres (Address), Yaş (Age) |
| Czech         | Adresa (Address), Číslo (Number) |
| Polish        | Ulica (Street), Wiek (Age) |
| Hungarian     | Cím (Address), Születésnap (Birthday) |
| Greek         | Διεύθυνση (Address), Ηλικία (Age) |
| Korean        | 주소 (Address), 전화번호 (Phone Number) |

andersonjeccel committed:
feat: Automatically generate form field values

Co-Authored-By: andersonjeccel <andersonjeccel@example.com>

andersonjeccel commented:
@RCheesley Fixed and covered by tests!

All your examples:

```
            ['My Option', 'my_option'],
            ['First Choice!', 'first_choice'],
            ['Test-Value_123', 'test_value_123'],
            ['Special@#$%Characters', 'special_characters'],
            ['  Trimmed  Spaces  ', 'trimmed_spaces'],
            ['Multiple___Underscores', 'multiple_underscores'],
            ['Àccénted Chàracters', 'accented_characters'],
            ['Привет мир', 'privet_mir'],
            ['', ''],
            ['123Numbers', '123numbers'],
            ['Numéro', 'numero'],
            ['Âge', 'age'],
            ['Straße', 'strasse'],
            ['Über', 'uber'],
            ['Teléfono', 'telefono'],
            ['Dirección', 'direccion'],
            ['Coração', 'coracao'],
            ['Ação', 'acao'],
            ['Födelsedag', 'fodelsedag'],
            ['Äktenskap', 'aktenskap'],
            ['Номер', 'nomer'],
            ['Адрес', 'adres'],
            ['电话', 'dian_hua'],
            ['地址', 'de_zhi'],
            ['電話番号', 'dian_hua_fan_hao'],
            ['住所', 'zhu_suo'],
            ['رقم', 'rqm'],
            ['عنوان', 'nwan'],
            ['Adres', 'adres'],
            ['Yaş', 'yas'],
            ['Adresa', 'adresa'],
            ['Číslo', 'cislo'],
            ['Ulica', 'ulica'],
            ['Wiek', 'wiek'],
            ['Cím', 'cim'],
            ['Születésnap', 'szuletesnap'],
            ['Διεύθυνση', 'dieuthynse'],
            ['Ηλικία', 'elikia'],
            ['주소', 'juso'],
            ['전화번호', 'jeonhwabeonho'],
```

kingsedem reviewed (approved):
Tested and works as expected

<img width="1695" height="822" alt="image" src="https://github.com/user-attachments/assets/76387479-9911-4767-bb14-a616d74c9578" />

kuzmany commented on app/bundles/CoreBundle/Form/Type/SortableValueLabelListType.php:
```diff
@@ -45,4 +63,21 @@ public function buildView(FormView $view, FormInterface $form, array $options):
+    /**
+     * Convert a string to a URL-safe slug.
+     */
+    private function slugify(string $text): string
```
I would choose already exists method for it https://github.com/mautic/mautic/blob/7.x/app/bundles/CoreBundle/Helper/InputHelper.php#L485

andersonjeccel replied:
Done!

andersonjeccel committed:
Merge branch '7.x' into UXUI-225-make-values-in-form-field-properties-get-automatically-generated

andersonjeccel committed:
Use existing method

kuzmany commented on app/bundles/CoreBundle/Form/Type/SortableValueLabelListType.php:
```diff
@@ -29,11 +32,27 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
+                $data = $event->getData();
+
+                if (is_array($data)
+                    && empty($data['value'])
```
Issue: empty() treats '0' as empty, so a user entering '0' as a value would have it auto-generated. This may not be desired behavior.

  Suggestion: Use strict empty check:
  if (is_array($data)
      && (!isset($data['value']) || $data['value'] === '')
      && !empty($data['label'])
  ) {

andersonjeccel replied:
Done!

kuzmany commented on app/bundles/CoreBundle/Form/Type/SortableValueLabelListType.php:
```diff
@@ -29,11 +32,27 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
+                    && empty($data['value'])
+                    && !empty($data['label'])
+                ) {
+                    $data['value'] = strtolower(preg_replace('/_+/', '_', trim(InputHelper::alphanum(InputHelper::transliterate((string) $data['label']), false, '_'), '_')));
```
This is hard to read and debug. Consider extracting to a private method:

  private function generateSlugFromLabel(string $label): string
  {
      $transliterated = InputHelper::transliterate($label);
      $alphanumeric = InputHelper::alphanum($transliterated, false, '_');
      $trimmed = trim($alphanumeric, '_');
      $normalized = preg_replace('/_+/', '_', $trimmed);
      return strtolower($normalized);
  }

Or use only already created and tested method, even If we do not cover all situations.

andersonjeccel replied:
Also done, thank you!

kuzmany reviewed (changes requested):
Good job.
I see one issue and one improvements for consideration.

andersonjeccel committed:
Address code review

andersonjeccel committed:
Improve test reliability

andersonjeccel committed:
csfix

andersonjeccel committed:
Merge branch '7.x' into UXUI-225-make-values-in-form-field-properties-get-automatically-generated

andersonjeccel committed:
rector

andersonjeccel committed:
Merge branch '7.x' into UXUI-225-make-values-in-form-field-properties-get-automatically-generated

andersonjeccel committed:
Merge branch '7.x' into UXUI-225-make-values-in-form-field-properties-get-automatically-generated

nileshlohar commented:
@andersonjeccel - Can you please check the test build failures.

andersonjeccel committed:
Merge branch '7.x' into UXUI-225-make-values-in-form-field-properties-get-automatically-generated

escopecz commented on app/bundles/CoreBundle/Form/Type/SortableValueLabelListType.php:
```diff
@@ -45,4 +64,14 @@ public function buildView(FormView $view, FormInterface $form, array $options):
         $view->vars['preaddon']      = $options['attr']['preaddon'] ?? [];
         $view->vars['postaddon']     = $options['attr']['postaddon'] ?? [];
     }
+
+    private function generateSlugFromLabel(string $label): string
+    {
+        $transliterated = InputHelper::transliterate($label);
+        $alphanumeric   = InputHelper::alphanum($transliterated, false, '_');
+        $trimmed        = trim($alphanumeric, '_');
+        $normalized     = preg_replace('/_+/', '_', $trimmed);
+
+        return strtolower($normalized ?? '');
+    }
```
Any reason why https://github.com/mautic/mautic/blob/915b49c5dee5b905e7f8b8c831860de08a55731f/app/bundles/CoreBundle/Helper/UrlHelper.php#L299 couldn't be used?

andersonjeccel replied:
The logic is intentionally different

- It transliterates first and then forces ASCII alphanumerics plus underscores, and guarantees a safe token even when the label is non‑Latin
- uses underscores as separators and normalizes consecutive underscores, never hyphens and or Unicode characters
- returns empty if transliteration fails, while stringURLUnicodeSlug never indicates failure and may return Unicode output

So using it here would change behavior

_(do you remember that bug when campaign wasn't able to recognize that the custom field value was equal to the value stored in database because they were sanitized causing mismatch? it'd be similar here)_.

escopecz commented on app/bundles/CoreBundle/Tests/Unit/Form/Type/SortableValueLabelListTypeTest.php:
```diff
@@ -0,0 +1,251 @@
+<?php
+
```
```suggestion

declare(strict_types=1);

```

escopecz commented on app/bundles/CoreBundle/Tests/Unit/Form/Type/SortableValueLabelListTypeTest.php:
```diff
@@ -0,0 +1,251 @@
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\Form\FormView;
+
+class SortableValueLabelListTypeTest extends TestCase
```
```suggestion
final class SortableValueLabelListTypeTest extends TestCase
```

kuzmany committed:
Add safety check for transliterate() returning false

InputHelper::transliterate() can return false on failure. Add an early
return to handle this edge case gracefully.

andersonjeccel committed:
Update app/bundles/CoreBundle/Tests/Unit/Form/Type/SortableValueLabelListTypeTest.php

Co-authored-by: John Linhart <jan@linhart.email>

andersonjeccel committed:
Update app/bundles/CoreBundle/Tests/Unit/Form/Type/SortableValueLabelListTypeTest.php

Co-authored-by: John Linhart <jan@linhart.email>

andersonjeccel committed:
Merge branch '7.x' into UXUI-225-make-values-in-form-field-properties-get-automatically-generated

andersonjeccel committed:
fix test

andersonjeccel committed:
Merge branch '7.x' into UXUI-225-make-values-in-form-field-properties-get-automatically-generated

andersonjeccel committed:
Merge branch '7.x' into UXUI-225-make-values-in-form-field-properties-get-automatically-generated

escopecz reviewed (approved):
I see no issue in the code changes 👍
